pkg/registry/resolver.go

```diff
@@ -323,6 +333,10 @@ func (r *registry) processStep(step *api.TestStep, seen sets.Set[string], stack
 		}
 		ret.Dependencies = deps
 	}
+
+	if ret.DNSConfig != nil {
```

Comment:
OK so going through case by case:

Currently, resolve DNS always grabs the highest-order set-DNS, and setIfUnset only pushes configuration "down" the stack, so it's actually somewhat useless in its current implementation.

If I write a job and want to reset DNSConfig (i.e. empty it), I could overwrite it using (1). This is currently being inherited by all sub-steps. As you mentioned, this is the same as setting it to something explicitly. (3)

If I write a job and want to inherit the highest-order DNSConfig, I could not set my DNSconfig, or set it to nil, case (2).

The recursive case is an interesting one. This is where the analog of using environment variable overrides breaks down a little. For an environment variable, you only have 2 options:
1. overwrite
2. inherit

DNSConfig could be broken down into 3 options:
1. overwrite
2. inherit
3. merge

Merge can potentially also be broken down into merge-append and merge-prepend.

I would argue that the simplest and safest way to implement merge is to not implement merge. Instead, we treat these like environment variables and only allow users to set it explicitly or inherit it completely.

My reasoning for this is that if you have the higher level of knowledge to write a parent workflow or job, you have all the information you need to explicitly write a merged DNSConfig of any variety you want. (Prepended contents, appended contents, overwritten contents, etc).

Here is the only case that doesn't work:
Say I write a step and want to make sure that that step's DNSConfig cannot be overwritten. Currently there is no way to do that. The only way to ensure that the step's DNSConfig is relevant is to make sure all the higher level objects are unset(/nil).

Is this a use case we want to support? This is unclear to me. The only major downside of the current approach (IMHO) is that if you have a complex DNS setup that needs to merge the needs across the whole setup, this will need to be set at a high level which means you could theoretically end up with conflicts if you have 2 steps that need to resolve the same name to different places.

Reply:
> Merge can potentially also be broken down into merge-append and merge-prepend.

Yes, you're right. Supporting a merge strategy requires us to implement one or more of the following (at least):
1. make a choice between merge-append or merge-prepend and I'm pretty sure that none of them is flexible enough to cover every use case.
2. enrich the `dnsConfig` stanza even more and allow the user to decide what strategy has to be used: `append`, `prepend`, etc. Here how it would look like:
```yaml
dnsConfig:
  nameserver: [ ... ]
  searches: [ ... ]
  merge_criteria: append
```

(2) is better than (1) but it's more complex and as of now I don't see any real need that would justify adding that complexity, therefore:

> I would argue that the simplest and safest way to implement merge is to not implement merge